File: ScapesCore/src/main/java/org/tobi29/scapes/client/gui/desktop/GuiComponentCredits.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.client.gui.desktop;

import org.tobi29.scapes.engine.gui.GuiComponentHeavy;
import org.tobi29.scapes.engine.gui.GuiLayoutData;
import org.tobi29.scapes.engine.gui.GuiRenderer;
import org.tobi29.scapes.engine.graphics.FontRenderer;
import org.tobi29.scapes.engine.graphics.GL;
import org.tobi29.scapes.engine.graphics.Matrix;
import org.tobi29.scapes.engine.graphics.Shader;
import org.tobi29.scapes.engine.utils.math.vector.Vector2;

public class GuiComponentCredits extends GuiComponentHeavy {
    private final String text;
    private float y = 160.0f;

    public GuiComponentCredits(GuiLayoutData parent, String text) {
        super(parent);
        this.text = text;
    }

    @Override
    protected void updateMesh(GuiRenderer renderer, Vector2 size) {
        FontRenderer font = gui.style().font();
        font.render(FontRenderer.to(renderer, 1.0f, 1.0f, 1.0f, 1.0f), text,
                size.floatY(), size.floatX());
    }

    @Override
    protected void transform(Matrix matrix, Vector2 size) {
        matrix.translate(0, y, 0);
    }

    @Override
    public void renderComponent(GL gl, Shader shader, Vector2 size,
            double delta) {
        y -= 40.0f * delta;
    }
}

File: ScapesFrontend/src/main/kotlin/org/tobi29/scapes/client/gui/GuiComponentCredits.kt
/*
 * Copyright 2012-2018 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.client.gui

import org.tobi29.math.vector.Vector2d
import org.tobi29.scapes.engine.graphics.FontRenderer
import org.tobi29.scapes.engine.graphics.GL
import org.tobi29.scapes.engine.graphics.Matrix
import org.tobi29.scapes.engine.graphics.Shader
import org.tobi29.scapes.engine.gui.GuiComponentHeavy
import org.tobi29.scapes.engine.gui.GuiLayoutData
import org.tobi29.scapes.engine.gui.GuiRenderer

class GuiComponentCredits(parent: GuiLayoutData) : GuiComponentHeavy(
        parent) {
    private var y = 160.0f
    var text: String = ""
        set(value) {
            if (field != value) {
                field = value
                dirty()
            }
        }

    override fun updateMesh(renderer: GuiRenderer,
                            size: Vector2d) {
        val font = gui.style.font
        font.render(FontRenderer.to(renderer, 1.0, 1.0, 1.0, 1.0), text, size.y,
                size.x)
    }

    override fun transform(matrix: Matrix,
                           size: Vector2d) {
        matrix.translate(0f, y, 0f)
    }

    public override fun renderComponent(gl: GL,
                                        shader: Shader,
                                        size: Vector2d,
                                        pixelSize: Vector2d,
                                        delta: Double) {
        y -= (40.0f * delta).toFloat()
    }
}
